### ISSUE TITLE
“unamed prepared statement” for postgresql13

### ISSUE THREAD

**Issue by yghenu** [knex/knex#5942]
when run code like this:
db('activity_nodes').select(['name as test_v1_main_5','path','parent_id','creator']).whereRaw(`name like  ?`,['%课程%'])
the PG recieved "unamed prepared statement ", so the select send to master pg, but i want send to readonly slave throw Database agent；
how can i use knex do some select opration by normal prepare statment or just excute?

**Comment by mercmobily** [comment]
Hi, thanks for reporting this.

PostgreSQL uses the extended query protocol by default for parameterised queries, which shows up as an “unnamed prepared statement” in server logs. You can force simple query mode per statement in Knex to avoid that and keep read-only routing that depends on simple queries.

Knex attaches per-query `options` to the driver request (lib/query/querycompiler.js:77-85) and the PostgreSQL client merges those into `queryConfig` (lib/dialects/postgres/index.js:170-209). The `pg` driver supports `{ simple: true }` to use the simple protocol and not create an unnamed prepared statement. You can set it like this:

```
// example based on your snippet
await db('activity_nodes')
  .select(['name as test_v1_main_5', 'path', 'parent_id', 'creator'])
  .whereRaw('name like ?', ['%课程%'])
  .options({ simple: true });
```

The `.options()` builder API is available on all queries and carries through to the driver (lib/builder-interface-augmenter.js:50-56 and lib/query/querycompiler.js:77-85). With `simple: true`, the DB agent should no longer see an unnamed prepared statement and can route the read to the replica.

If you need named prepared statements instead, Knex does not currently assign `name` automatically on pg requests; supporting named statements would require a feature addition to populate `queryConfig.name` predictably.

In conclusion, use `.options({ simple: true })` on pg queries that must avoid the extended protocol. If that does not satisfy your routing, please share a minimal repro and your pg versions.

Since this issue is so old, I will close it for now.
